### PR TITLE
feat: redeem for collection holder

### PR DESCRIPTION
**Pull request by c98tristan**
- [x] Implement function
- [x] Unit test happy case
- [x] Unit test cover

### CODE REVIEW

**Review by imterryyy98 — CHANGES_REQUESTED** [review]

File: contracts/MatrixVault.sol [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+
+import "./Coin98VaultV2.sol";
+import "./lib/Token.sol";
+import "./lib/VRC725.sol";
+
+contract MatrixVault is Coin98VaultV2 {
+    using AdvancedERC20 for IERC20;
+
+    // NFT ID => (Event ID => Claimed)
+    mapping(uint256 => mapping(uint256 => bool)) internal _isClaimed;
```
Comment (imduchuyyy) [inline contracts/MatrixVault.sol:12]: Rename to `isClaimedByNFT`

File: contracts/MatrixVault.sol [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+        bytes32[] calldata proofs
+    ) public payable {
+        uint256 fee = IVaultConfig(_factory).fee();
+        uint256 gasLimit = IVaultConfig(_factory).gasLimit();
+        if (fee > 0) {
+            require(_msgValue() == fee, "C98Vault: Invalid fee");
+        }
+        require(collectionAddress != address(0), "C98Vault: Invalid collection");
+        require(receiver != address(0), "C98Vault: Invalid receiver");
+        require(VRC725(collectionAddress).balanceOf(receiver) > 0, "C98Vault: Receiver has no NFT");
```
Comment (imduchuyyy) [inline contracts/MatrixVault.sol:47]: We can remove this check safely because we already have check for owner of TokenId should equal to receiver

**Review by imduchuyyy — CHANGES_REQUESTED** [review]

File: contracts/MatrixVault.sol (hunk 1 above)
Comment (imterryyy98) [inline contracts/MatrixVault.sol:12]: rename to `_isClaimedByToken`

**Review by imterryyy98 — CHANGES_REQUESTED** [review]

**Comment by imterryyy98** [comment]
I suggest to split `redeemForCollectionHolder` into 2 functions, one function for specific tokenId, the other one for all tokenIds. You can implement internal functions to handle overlap validation logic and execution.

**Review by imterryyy98 — APPROVED** [review]
LGTM

**Pull request merged**